### ISSUE TITLE
[Proposal] Consistent `batch_first` effect for RNN modules

### ISSUE THREAD

**Issue by Stonesjtu** [pytorch/pytorch#4145]
In current RNN (including `nn.RNN`, `nn.GRU`, `nn.LSTM`) implementation, the `batch_first` will only affect the dimension order of **input** and **output** variables. But there are also **hidden states** returned from RNN, and the *batch* dimension is always placed in the 2nd dimension regardless of the `batch_first` argument.
Could someone tell me the advantages of such odd behaviour? Or I can help to patch this.

**Comment by ngimel** [comment]
Without batch_first, input dimensions are [seq_length, batch_size, num_features]. batch_first switches the seq_length and batch_size for input. Hidden dimension is [num_layers * num_directions, batch, hidden_size], it is for the last step in a sequence, so seq_length dimension is squeezed out for hidden. Since semantically hidden dimensions are different from input dimensions, it is unclear why batch_first switch should affect them.

**Comment by Stonesjtu** [comment]
Yes, the hidden and input dimensions are different in RNN, but they both share the `batch_size` dimension. For any NN structures, `batch` should be an independent dimension.
I don't understand why it becomes unclear just because the other two dimensions differ.

**Comment by iqDF** [comment]
Sorry for arriving late. I would agree with @Stonesjtu . Here having the dimensions ordered the same for both output of the RNN, namely `output` and `h_n` are more convenient for developers. So if *input* and *output* has dimension of `(batch_size, seq_length, num_features)` then so shall be the hidden states `h_n` to have `batch_size` as first dimension for consistency.

Thus, I would like to point out that the documentations only says that `batch_size=1` will make *input* and *output* to have `batch_size` as first dimension. This is ambiguous as hidden state `h_n` is also considered as *output* if you think of it.

**Comment by zou3519** [comment]
I think the best thing to do here is to clarify the documentation on what `batch_first` does. Many people get confused about the hidden states not having their batch first.

I agree with @ngimel that it is unclear from a semantic perspective why `batch_first` should affect the hidden states (because the functionality of batch_first is to switch the timestep dimension and the batch dimension and there is no timestep dimension in the hidden state).

**Comment by ezyang** [comment]
`swap_batch_and_timestep`? In the triage meeting, we agreed to modify the documentation / change the naming so that it is clearer (but not chnage the actual order.)

**Comment by ngimel** [comment]
`swap_batch_and_timestep=True/False` does not say which one becomes first. `format=batch_time_features/time_batch_features'? Named tensors, please save us from this!

**Comment by clthorre** [comment]
Just commenting to also express confusion at having set batch_first=True and then having the output which I'm using, h_n, not have the first dimension as batch. 

Maybe even specifically saying in the docs it does not impact h_n, c_n. Because it currently says it changed input and output tensors, which causes confusion because it changes the tensors named "input" and "output" but there are more output tensors than just "output".

**Comment by sohaib023** [comment]
Just wanted to share my thoughts on this.
I find this flag (for altering input/output format) to be highly anomalous as compared to other layers in PyTorch and something one does not expect.
Imagine if Conv2d had an option to switch the channel dimension to make it consistent with (h, w, c) format of tensorflow. It would only cause confusions on open-source and collaborative projects. 
As I'm not 100% sure on the need of this flags existence so I wont comment on that, but at the very least the documentation should have explicit up-front warnings or notes about this behavior. For someone who's dropping in to the documentation for just checking the format will look at following line and will be scratching his head for hours over such a simple bug:
![image](https://user-images.githubusercontent.com/21100293/102539027-ec7d3000-40ce-11eb-8167-2770407b8e49.png)

**Comment by zou3519** [comment]
Raising priority based on user activity. I am not sure we can change the behavior without doing a deprecation cycle, but putting a warning into the documentation is simple enough

**Comment by VitalyFedyunin** [comment]
Hi-pri to fix docs.

**Comment by jbschlosser** [comment]
Removing high-pri now that the docs have been updated.
